module/platform/Webots/src/Webots.cpp

```diff
@@ -0,0 +1,228 @@
+/*
+ * This file is part of the NUbots Codebase.
+ *
+ * The NUbots Codebase is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * The NUbots Codebase is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with the NUbots Codebase.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Copyright 2021 NUbots <nubots@nubots.net>
+ */
+
+#include "Webots.hpp"
+
+#include <fmt/format.h>
+
+#include "extension/Configuration.hpp"
+
+#include "message/motion/ServoTarget.hpp"
+#include "message/output/CompressedImage.hpp"
+#include "message/platform/darwin/DarwinSensors.hpp"
+#include "message/platform/webots/ConnectRequest.hpp"
+#include "message/platform/webots/messages.hpp"
+
+// Include headers needed for TCP connection
+#include <netdb.h>      /* definition of gethostbyname */
+#include <netinet/in.h> /* definition of struct sockaddr_in */
+#include <sys/socket.h>
+#include <sys/time.h>
+#include <unistd.h> /* definition of close */
+
+namespace module::platform {
+
+using extension::Configuration;
+using message::motion::ServoTargets;
+using message::output::CompressedImage;
+using message::platform::darwin::DarwinSensors;
+
+using message::platform::webots::ActuatorRequests;
+using message::platform::webots::ConnectRequest;
+using message::platform::webots::MotorPosition;
+using message::platform::webots::MotorTorque;
+using message::platform::webots::MotorVelocity;
+using message::platform::webots::SensorMeasurements;
+
+int Webots::tcpip_connect(const std::string& server_name, const int& port) {
+    // Create the socket
+    int fd = socket(AF_INET, SOCK_STREAM, 0);
+
+    // Store the ip address information that we will connect to
+    sockaddr_in address;
+    memset(&address, 0, sizeof(sockaddr_in));
+    address.sin_family = AF_INET;
+    address.sin_port   = htons(port);
+    hostent* server    = gethostbyname(server_name.c_str());
+
+    // Check that dns was successful
+    if (server) {
+        std::memcpy(reinterpret_cast<char*>(&address.sin_addr.s_addr),
+                    reinterpret_cast<char*>(server->h_addr),
+                    server->h_length);
+    }
+    else {
+        close(fd);
+        log<NUClear::ERROR>(fmt::format("Cannot resolve server name: {}", server_name));
+        return -1;
+    }
+
+    // Connect to the ip address
+    if (int error = connect(fd, (struct sockaddr*) &address, sizeof(address))) {
```

Comment:
I want to note that I saw that everyone seems to cast their `sockaddr_in` like this when calling `bind(..)`, but I couldn't see people doing it for `connect(..)`, because they all seemed to do things fundamentally differently, usually using `getaddrinfo(..)` instead of `gethostbyname(..)`

Reply:
`sockaddr*` can be considered a "base class" as much as c has such a concept. Beyond that there are variants of `sockaddr` for each of the different network types.

For example `sockaddr_in` is the IPv4 version of the struct. and `sockaddr_in6` would be the IPv6 version. There are also other versions for other socket protocols that can be used (although those two are the most common you'll see)

All the functions require you to cast your type to a `sockaddr*` message to use, and to make sure that the size you pass in is actually the size of the type itself (`sizeof(sockaddr)` might not equal `sizeof(sockaddr_in)`. If you ever need to store a list of addresses that contain a mix of those there is a special type `sockaddr_storage` which will always have a large enough size to store any of the possible `sockaddr*` variants.

In this case though, you should use a reinterpret cast, not a c style cast

Reply:
Thanks for the update! that's quite interesting. 
One thing I did find was that `sizeof(sockaddr) == sizeof(sockaddr_in)`, which is convenient for what we're doing

Reply:
`sockaddr` will be the same size as sockaddr_in mostly because of people using the wrong one all the time and sockaddr_in being the most common :P. Since using sizeof(sockaddr) is so common instead of sizeof(sockaddr_in) for people they just made them the same size by padding out the `sockaddr` message with reserved bytes :P